### ISSUE TITLE
FlexAttention vs. Torch.sdpa

### ISSUE THREAD

**Issue by rwn17** [co-me-tokens/CoMe#5]
Hi @MarkChenYutian ,

Thanks for your excellent work! I'm curious about the accelerating ratio of VGGT* shown in Figure 9, where FlexAttention is about 4 times faster than torch sdpa. I tested it with your benchmarking code [https://github.com/co-me-tokens/CoMe/blob/main/Accelerate/common.py](https://github.com/co-me-tokens/CoMe/blob/main/Accelerate/common.py). Here is the result without attn_bias. 

```
<torch.utils.benchmark.utils.common.Measurement object at 0x1503f59431c0>
SDPA Backend = torch-flash: B=8 H=16 C=64 N=2048, attn_bias=False, dtype=fp16
  475.05 us
  1 measurement, 100 runs , 12 threads
<torch.utils.benchmark.utils.common.Measurement object at 0x1503f4567130>
SDPA Backend = torch-flex: B=8 H=16 C=64 N=2048, attn_bias=False, dtype=fp16
  522.32 us
  1 measurement, 100 runs , 12 threads

<torch.utils.benchmark.utils.common.Measurement object at 0x1503f4567e20>
SDPA Backend = torch-flash: B=1 H=16 C=64 N=16384, attn_bias=False, dtype=fp16
  3.59 ms
  1 measurement, 100 runs , 12 threads
<torch.utils.benchmark.utils.common.Measurement object at 0x1503f4574b20>
SDPA Backend = torch-flex: B=1 H=16 C=64 N=16384, attn_bias=False, dtype=fp16
  3.78 ms
  1 measurement, 100 runs , 12 threads
```

I tested this on PyTorch 2.6.1 using H100, H800, and 4090 GPUs, and FlexAttention is consistently slower than sdpa. Could you provide any idea about the accelerating ratio from FlexAttention in Figure 9? Thanks in advance.

Bests,
Weining

**Comment by rwn17** [comment]
Hi @MarkChenYutian ,

Thanks for your prompt reply. I checked the [source code](https://github.com/pytorch/pytorch/blob/c260ae19a70c72de6f8cff656cd46db5572aedc3/aten/src/ATen/native/transformers/cuda/sdp_utils.cpp#L104-L116) of the sdpa backend selection. For the default case of VGGT global attention (without mask) using CUDA, the order would be cudnn ->flash->mem-efficient->math.

**Comment by MarkChenYutian** [comment]
Hi @rwn17 , thanks for digging into this - I also dig into the pytorch source code and see that the kernel priority seems to be actively changed in the recent releases.

`sdp_utils.cpp` is changed from ver 2.8 to latest commit you referred to above: 
- https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/aten/src/ATen/native/transformers/cuda/sdp_utils.cpp#L80-L83

Here seems to be the ver 2.8 pytorch's sdp priority definition:
- https://github.com/pytorch/pytorch/blob/ba56102387ef21a3b04b357e5b183d48f0afefc7/aten/src/ATen/Context.h#L423-L427

On this (release 2.8) branch, the priority seems to be flash -> mem-efficient -> math -> cudnn.

When I experimented on the cluster this August, I believe the default kernel was `mem-efficient` for some reason... So the acceleration ratio from VGGT to VGGT* is from `mem-efficient` SDPA to `flexattention` SDPA

**Comment by rwn17** [comment]
Hi @MarkChenYutian ,

Thank you for pointing out the change in the default backend. I’ve also noted that the SDPA Flash Attention backend requires several checks to pass—specifically regarding the cuDNN version, hardware compatibility, and data types (it only supports Float16 and BFloat16). If these requirements aren't met, the system simply falls back to a different implementation. My concerns have been addressed, so I will close this issue now.
